### ISSUE TITLE
Multiple requests with CustomSessionDelegate, sessionDidReceiveChallengeWithCompletion called only once!

### ISSUE THREAD

**Issue by goktugaral** [Alamofire/Alamofire#2493]
Hi,
Sorry for did not fill the template.

I make multiple requests which have my customDelegate
Here is;

```
self.sessionManager = SessionManager(delegate: (oauthType == .noSecured) ? SessionDelegate() : FingerprintSessionDelegate())
self.sessionManager.session.configuration.timeoutIntervalForRequest = ServiceConstants.kRequestTimeout
self.sessionManager.request(encodedurl, headers: headers).responseJSON { [weak self] response in
                
  self?.requestFinished(response: response, callback: { (result) in
         callback(result)
   }, errorCallback: { (serviceResult) in
          errorCallback!(serviceResult)
   })
}
```

```
class FingerprintSessionDelegate: SessionDelegate {
       override init() {
               super.init()
        
               // Alamofire uses a block var here
               sessionDidReceiveChallengeWithCompletion = { session, challenge, completion in

               let storedFingerprint = CoreManager.instance.getFingerprint(key: "SHA256")?.decomposeFingerprint()
  
              let fingerprint = challenge.getSHA256Fingerprint()
              if storedFingerprint == fingerprint {
                  completion(.performDefaultHandling,nil)
              }else{
                  completion(.cancelAuthenticationChallenge,nil)
              }
          }
      }
}
```

If I call request like this;
```
self.requestSettings()
self.requestTermsAndConditions()

```
my customDelegate's sessionDidReceiveChallengeWithCompletion called only once.

But if I call nested, it works;
```
self.requestSettings() {
     success {
           self.requestTermsAndConditions()
     }
}
```

What is the reason of this? What do I miss?

**Alamofire Environment**
Alamofire version: 4.7.1
Xcode version: 9.3
Platform(s) running Alamofire: Using Simulator
macOS version running Xcode: 10.13.4

**Comment by goktugaral** [comment]
I discovered another thing;
I make multiple requests which have my customDelegate but the thing is it did not call sessionDidReceiveChallengeWithCompletion twice...Why is that?

```
override init() {
        super.init()
        
        // Alamofire uses a block var here
        sessionDidReceiveChallengeWithCompletion = { session, challenge, completion in

            let storedFingerprint = CoreManager.instance.getFingerprint(key: "SHA256")?.decomposeFingerprint()
  
            let fingerprint = challenge.getSHA256Fingerprint()
            if storedFingerprint == fingerprint {
                completion(.performDefaultHandling,nil)
            }else{
                completion(.cancelAuthenticationChallenge,nil)
            }
        }
    }
```

**Comment by jshier** [comment]
Unfortunately, there's not enough context here for us to be able to tell what's going on.

First, I can tell you that it's not recommended to have separate `SessionManager`s per request. Instead you should keep a reference to a single (or few) `SessionManager` that you use for every request. But I wouldn't recommend it anyway.

Second, the `sessionDidReceiveChallengeWithCompletion` closure is called in the `URLSessionDelegate`'s `urlSession(_ session:, didReceive challenge:, completionHandler:)` method. According to Apple's documentation, this method is called in two situations:

- When a remote server asks for client certificates or Windows NT LAN Manager (NTLM) authentication, to allow your app to provide appropriate credentials
- When a session first establishes a connection to a remote server that uses SSL or TLS, to allow your app to verify the server’s certificate chain

Which means that you may only receive it once per sequence of requests, as your connection(s) may be reused, which means you won't see an additional challenge.

Third, and not related to you issue per se, is that setting `self.sessionManager.session.configuration.timeoutIntervalForRequest` won't work, as changes to the `URLSessionConfiguration` don't have any effect after it has been added to a `URLSession`. You'll want to create a configuration and a session and use that session to initialize the `SessionManager`.

Ultimately, we can only help if you are able to provide more context about your requirements, which is what the issue template is designed to provide.

**Comment by jshier** [comment]
Please open a new issue, following the template, if you have new information. Otherwise StackOverflow may give you better results.
